identity-android/src/main/java/com/android/identity/android/direct_access/DirectAccess.kt

```diff
@@ -0,0 +1,493 @@
+import org.bouncycastle.asn1.ASN1UTCTime
+import java.io.ByteArrayOutputStream
+import java.io.IOException
+import java.nio.ByteBuffer
+import kotlin.time.Duration
+
+/**
+ * A class which handles common operations related to the DirectAccess applet.
+ */
+@RequiresApi(Build.VERSION_CODES.P)
+object DirectAccess {
+
+    private const val TAG = "DirectAccess"
+    private const val PROVISION_BEGIN: Byte = 0
+    private const val PROVISION_UPDATE: Byte = 1
+    private const val PROVISION_FINISH: Byte = 2
+    private const val KEY_CERT: Int = 0x01
+    private const val KEY_ENC_DATA: Int = 0x00
+    private const val CMD_MDOC_CREATE = 0x01
+    private const val CMD_MDOC_CREATE_PRESENTATION_PKG = 0x07
+    private const val CMD_MDOC_DELETE_CREDENTIAL = 0x08
+    private const val CMD_MDOC_PROVISION_DATA = 0x09
+    private const val CMD_MDOC_SWAP_IN = 0x06
+    private const val APDU_RESPONSE_STATUS_OK = 0x9000
+    private const val INS_ENVELOPE = 0xC3.toByte()
+
+    val transport = DirectAccessOmapiTransport()
+
+    /**
+     * Returns the maximum size, in bytes, of `credentialData` which can be used in credentials.
+     * This will be at least 32 KiB.
+     */
+    val maximumCredentialSize: Long
+        get() {
+            // todo integrate with applet
+            return 32768
+        }
+
+    /**
+     * Returns true if direct access is supported on the device, false otherwise.
+     */
+    val isDirectAccessSupported: Boolean
+        get() {
+            try {
+                transport.openConnection()
+            } catch (e: Throwable) {
+                Logger.d(TAG, "${e.message}")
+                return false
+            }
+            transport.closeConnection()
+            return true
+        }
+
+    private fun setShort(buf: ByteArray, offset: Int, value: Short) {
+        buf[offset] = (value.toInt() shr 8 and 0xFF).toByte()
+        buf[offset + 1] = (value.toInt() and 0xFF).toByte()
+    }
+
+    private fun getAPDUResponseStatus(input: ByteArray): Int {
+        // Last two bytes are the status SW0SW1
+        val SW0 = input[input.size - 2]
+        val SW1 = input[input.size - 1]
+        return SW0.toInt() shl 8 or SW1.toInt() and 0xFFFF
+    }
+
+    @Throws(IOException::class)
+    private fun makeCommandApdu(data: ByteArray): ByteArray {
+        // TODO Handle non extended length.
+        val bos = ByteArrayOutputStream()
+        bos.write(0) // CLS
+        bos.write(INS_ENVELOPE.toInt()) // INS
+        bos.write(0) // P1
+        bos.write(0) // P2
+        // Send extended length APDU always as response size is not known to HAL.
+        // Case 1: Lc > 0  CLS | INS | P1 | P2 | 00 | 2 bytes of Lc | CommandData | 2 bytes of Le
+        // all set to 00.
+        // Case 2: Lc = 0  CLS | INS | P1 | P2 | 3 bytes of Le all set to 00.
+        bos.write(0x00)
+        // Extended length 3 bytes, starts with 0x00
+        if (data.isNotEmpty()) {
+            bos.write(data.size shr 8)
```

Comment:
Check that data.size <= 0xffff, or this could cause problems.

Reply:
certify() and setActiveCredential() are the only two functions where there is a possibility of getting large data. These functions already handle sending large input data in chunks of smaller sizes to the applet. May I know if it is still required to handle this check here.

Reply:
Just to be on the safe side, or in case some of the surrounding code changes, a check() here would catch errors that could otherwise be transparent. So yeah, please add a check, even though the current usage should be safe.

Reply:
Sure. Will do.